### PR TITLE
Add MPS SDPA workarounds for value head dim and bidirectional attention

### LINKED ISSUE #44554: [MPS] Upstream correctness issue in attention when value head dim differs from query
### System Info

There is a correctness issue with attention in PyTorch when using the MPS backend with value head dims different from the query head (see https://github.com/pytorch/pytorch/issues/176767).

Consider the following reproducer

```python
import torch
import torch.nn.functional as F

q = torch.rand(1, 1, 8, 4, device="mps")
k = torch.rand(1, 1, 8, 4, device="mps")
v = torch.rand(1, 1, 8, 2, device="mps")

y_mps1 = F.scaled_dot_product_attention(q, k, v)
y_mps2 = F.scaled_dot_product_attention(q, k, v)

print(f"{y_mps1-y_mps2 = }")

y_cpu1 = F.scaled_dot_product_attention(q.cpu(), k.cpu(), v.cpu())
y_cpu2 = F.scaled_dot_product_attention(q.cpu(), k.cpu(), v.cpu())

print(f"{y_cpu1-y_cpu2 = }")

print(f"{y_mps1.shape = }")
print(f"{y_cpu1.shape = }")

# Output:
# y_mps1-y_mps2 = tensor([[[[ 0.0000,  0.0000,  0.0000,  0.0000],
#           [ 0.0000,  0.0000,  0.0000,  0.0000],
#           [-0.0957, -0.1705, -0.1087, -0.1648],
#           [-0.1155, -0.1248, -0.0946, -0.1255],
#           [-0.1089, -0.1662, -0.1261, -0.1559],
#           [-0.0904, -0.1262, -0.0927, -0.1335],
#           [-0.1080, -0.1644, -0.1226, -0.1611],
#           [-0.0849, -0.1324, -0.0984, -0.1281]]]], device='mps:0')
# y_cpu1-y_cpu2 = tensor([[[[0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.],
#           [0., 0.]]]])
# y_mps1.shape = torch.Size([1, 1, 8, 4])
# y_cpu1.shape = torch.Size([1, 1, 8, 2])
```

The numeric mismatch above is due to uninitialized memory being part of the return tensor. Notice that there is also a shape mismatch from the expected result.

The issue has been fixed in PyTorch (https://github.com/pytorch/pytorch/pull/176843), but won't be released until PyTorch 2.12.

I don't know if different dims for query and value is a common use-case for Hugging Face models, but I wanted to notify you of the issue regardless. I'm happy to provide a workaround if you think this issue should be fixed in transformers. Potential workaround is to resize the output tensor and wrangle the strides, as the calculation itself is good for the first (expected) numels.

The issue is similar to #44247, which is another correctness issue in MPS SDPA needing workaroud for specific PyTorch versions.

### Who can help?

@vasqu @Cyrilvallez

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

See reproducer above. I'm not sure if this is an issue in transformers.

### Expected behavior

See reproducer above.

### PR DESCRIPTION
**Pull request by moktamd**
Adds `_apply_mps_fixes` in `sdpa_attention.py` to handle two upstream PyTorch MPS bugs:

1. **pytorch/pytorch#176767** (fixed in PyTorch 2.12): pads value tensor when `v_head_dim != q_head_dim` to avoid corrupted output. Affects DeepSeek models with MQA.

2. **pytorch/pytorch#174861** (fixed in PyTorch 2.11): forces a non-bool attention mask for non-causal, non-float32 attention to route through `sdpa_general_mps` instead of broken `sdpa_vector_2pass_mps`.

Both fixes are version-gated and will no-op once the upstream PyTorch fixes are available.

Fixes #44554
Fixes #44247

### CODE REVIEW

**Comment by hvaara** [comment]
@moktamd Great that you want to contribute! Unfortunately these issues are actively worked on already. It is clear from my discussion with the maintainers in #44247 and #44554, on top of me being the author of both issues. Please find another issue to work on that is not already taken. Thank you.

**Comment by moktamd** [comment]
Apologies, I missed that you were already working on this. Closing in your favor. Good luck with the fix!
